aten/src/ATen/native/mkldnn/Gelu.cpp

```diff
@@ -0,0 +1,80 @@
+#include <ATen/ATen.h>
+#include <ATen/NativeFunctions.h>
+#include <ATen/Config.h>
+
+
+#if !AT_MKLDNN_ENABLED()
+
+namespace at { namespace native {
+
+Tensor mkldnn_gelu(const Tensor& input) {
+  TORCH_CHECK(false, "mkldnn_gelu: ATen not compiled with MKLDNN support");
+}
+
+Tensor mkldnn_gelu_backward(const Tensor& grad_output, const Tensor& input) {
+  TORCH_CHECK(false, "mkldnn_gelu_backward: ATen not compiled with MKLDNN support");
+}
+
+}}
+
+#else // AT_MKLDNN_EBABLED
+
+#include <ATen/native/mkldnn/MKLDNNCommon.h>
+#include <ATen/native/mkldnn/Utils.h>
+#include <ATen/native/mkldnn/Gelu.h>
+
+namespace at { namespace native {
+
+Tensor _mkldnn_gelu(const Tensor& input, const Tensor& result) {
+  const ideep::tensor& x = itensor_from_tensor(input);
+  ideep::tensor y;
+  if (!input.is_mkldnn()) {
+    // deal with dense layout
+    result.resize_(input.sizes(), input.suggest_memory_format());
+    y = itensor_from_tensor(result);
+  }
+  ideep::eltwise_forward::compute(
+      x, y, ideep::algorithm::eltwise_gelu_erf, ideep::prop_kind::forward_training, /*alpha*/ 0.0);
+  if (!input.is_mkldnn()) {
+    return result;
+  }
+  return new_with_itensor_mkldnn(std::move(y), optTypeMetaToScalarType(input.options().dtype_opt()),
+                                 input.options().device_opt());
```

Comment:
This isn't kosher. When a `result` tensor is explicitly passed, you MUST copy the result of the computation into that result. You're not allowed to just turn around and just return some other tensor; if someone passed an `out=` argument to write into a slice of some large tensor, you'd better do that!

Reply:
I'd like to see some test coverage for this case.

Reply:
In this PR, both CPU path and MkldnnCPU path will go into Gelu's mkldnn implementation. In CPU path, an `input` and a `result` tensor are explicitly passed, while in MkldnnCPU path, only `input` is passed.
In order to reuse the implementation, we abstract out the core function `_mkldnn_gelu`. In CPU path, it will call `_mkldnn_cpu` directly. In MkldnnCPU path, since it only has `input`, it will create an empty tensor with input size as `result`. `result` here is unused. Need return the tensor from `new_with_itensor_mkldnn`, which is the same as other Mkldnn operator implementations.

Reply:
I'm not a fan of the code reuse by swiss cheesing the conditionals and relying on different invariants at each call site. If you directly inline `_mkldnn_gelu` into the structured kernel you get a very simple gelu call (you can drop the resizes, because structured kernels handle that for you):

```
TORCH_IMPL_FUNC(gelu_out_cpu) (
  const Tensor& self, const Tensor& result
) {
  GeluKernel(kCPU, *this);
  if (use_mkldnn(self)) {
    const ideep::tensor& x = itensor_from_tensor(self);
    ideep::tensor y = itensor_from_tensor(result);
    ideep::eltwise_forward::compute(
       x, y, ideep::algorithm::eltwise_gelu_erf, ideep::prop_kind::forward_training, /*alpha*/ 0.0);
  } else {
    GeluKernel(kCPU, *this);
  }
}
```

This seems much easier to understand than the current formulation. (Missing here is some ifdefs, easy enough to add back).

I also did the other call site inlined and it also seems better:

```
Tensor mkldnn_gelu(const Tensor& input) {
  if (input.scalar_type() == ScalarType::BFloat16) {
    TORCH_CHECK(mkldnn_bf16_device_check(),
        "mkldnn_gelu: bf16 path needs the cpu support avx512bw, avx512vl and avx512dq");
  }
  const ideep::tensor& x = itensor_from_tensor(input);
  ideep::tensor y;
  ideep::eltwise_forward::compute(
      x, y, ideep::algorithm::eltwise_gelu_erf, ideep::prop_kind::forward_training, /*alpha*/ 0.0);
  return new_with_itensor_mkldnn(std::move(y), optTypeMetaToScalarType(input.options().dtype_opt()),
                                 input.options().device_opt());
}
```

You get to skip the unnecessary result allocation (which is subsequently ignored).

Reply:
Done.